### DISCUSSION TITLE
Exception when trying to use SIMDJSON_VERBOSE_LOGGING

### DISCUSSION THREAD

**Discussion by drcdr** [simdjson/simdjson#1938] [General]
I'm trying to debug why parsing a certain JSON file was giving long times for a certain key (as seen by the profiler).  Single-stepping, I noticed that there was a logging feature, so I thought I'd try and use that.  I'm using Visual Studio 2022, Debug mode, simdjson 3.0.1.
- I enabled SIMDJSON_VERBOSE_LOGGING in simdjson.h.  It then crashed in log_line().
- I simplified to the following, and still got the same crash:
```
	auto cars_json = R"( [
		{ "make": "Toyota", "model": "Camry",  "year": 2018, "tire_pressure": [ 40.1, 39.9, 37.7, 40.4 ] },
		{ "make": "Kia",    "model": "Soul",   "year": 2012, "tire_pressure": [ 30.1, 31.0, 28.6, 28.7 ] },
		{ "make": "Toyota", "model": "Tercel", "year": 1999, "tire_pressure": [ 29.8, 30.0, 30.2, 30.5 ] }
	] )"_padded;
	static simdjson::ondemand::parser  parser;
	simdjson::ondemand::document doc = parser.iterate(cars_json);
```
I'm seeing that the first time log_line is called, index is -1, which is trying to reference memory before the beginning of the buffer (iter.token.buf).  The error occurs here, with i=0, and delta=-1 from the calling routine.
```
      auto current_structural = &buf[*index];
      for (int i=0;i<LOG_BUFFER_LEN;i++) {
        printf("%c", printable_char(current_structural[i]));
      }
```

Before entering an issue, I just wanted to understand if I'm using SIMDJSON_VERBOSE_LOGGING correctly or not.

**Comment by lemire** [comment]
Most of our users do not know about logging, so it is not a well tested feature (it requires a specific macro).

Nevertheless, the problem is real and you can see a possible fix:
https://github.com/simdjson/simdjson/pull/1939/files

I will soon release a version with a patch. You are probably able to patch it locally for the time being?

**Comment by drcdr** [comment]
Thanks for the quick response! This fix eliminates the crash.  This logging is certainly a very nice feature.
I used it to try and optimize my implementation.  I'd done several on-demand parsers before, always using the if-then-on-object's-key() approach (e.g. as done [here](https://github.com/simdjson/simdjson/issues/1378#issuecomment-756399930))

Looking at the [v2.0 on-demand writeup](https://simdjson.org/api/2.0.0/md_doc_ondemand_design.html) I decided to try the ["key"] approach instead, and this is where I'm now seeing lots of skips.

In a reproducible example:
```
	auto cars_json = R"( [
		{ "make": "Toyota", "data": {"model": "Camry",  "year":"2018", "rating":"A"}, "tire_pressure": [ 40.1, 39.9, 37.7, 40.4 ] },
		{ "make": "Kia",    "data": {"model": "Soul",   "year":"2012", "rating":"B"}, "tire_pressure": [ 30.1, 31.0, 28.6, 28.7 ] },
		{ "make": "Toyota", "data": {"model": "Tercel", "year":"1999", "rating":"A"}, "tire_pressure": [ 29.8, 30.0, 30.2, 30.5 ] }
	] )"_padded;
	static simdjson::ondemand::parser  parser;
	simdjson::ondemand::document doc = parser.iterate(cars_json);

	for (simdjson::ondemand::object o : doc)
	{
		printf("***** make *****\n");
		std::string_view make = o["make"];
		printf("***** data/model *****\n");
		std::string_view model = o["data"]["model"];
		printf("***** data/year *****\n");
		std::string_view year = o["data"]["year"];
		printf("***** data/rating *****\n");
		std::string_view rating = o["data"]["rating"];
		break;  // only need to look at first iteration
	}
```

My logging output shows that o["make"] is efficient, o["data"]["model"] is efficient, but the next two have skips (data/rating not shown):
```
| Event                          | Buffer                         | Next       | Depth | Detail |
|--------------------------------|--------------------------------|------------|-------|--------|
|   +document                    |                                | [   { "mak |     1 |        |
|   +array                       | [   { "make": "Toyota", "data" | { "make":  |     1 |        |
|     +object                    | { "make": "Toyota", "data": {" | "make": "T |     2 |        |
***** make *****
|       match                    | "make": "Toyota", "data": {"mo | : "Toyota" |     3 |   make |
|       string                   | "Toyota", "data": {"model": "C | , "data":  |     3 |        |
|       string                   | "Toyota", "data": {"model": "C | , "data":  |     3 |        |
***** data/model *****
|       match                    | "data": {"model": "Camry",  "y | : {"model" |     3 |   data |
|       +object                  | {"model": "Camry",  "year":"20 | "model": " |     3 |        |
|         match                  | "model": "Camry",  "year":"201 | : "Camry", |     4 |  model |
|         string                 | "Camry",  "year":"2018", "rati | ,  "year": |     4 |        |
|         string                 | "Camry",  "year":"2018", "rati | ,  "year": |     4 |        |
***** data/year *****
|       skip                     | ,  "year":"2018", "rating":"A" | "year":"20 |     3 |        |
|       skip                     | "year":"2018", "rating":"A"},  | :"2018", " |     3 |        |   <===== (1)
|       skip                     | :"2018", "rating":"A"}, "tire_ | "2018", "r |     3 |        |
|       skip                     | "2018", "rating":"A"}, "tire_p | , "rating" |     3 |        |
|       skip                     | , "rating":"A"}, "tire_pressur | "rating":" |     3 |        |
|       skip                     | "rating":"A"}, "tire_pressure" | :"A"}, "ti |     3 |        |
|       skip                     | :"A"}, "tire_pressure": [ 40.1 | "A"}, "tir |     3 |        |
|       skip                     | "A"}, "tire_pressure": [ 40.1, | }, "tire_p |     3 |        |
|       -skip                    | }, "tire_pressure": [ 40.1, 39 | , "tire_pr |     3 |        |
|       no match                 | "tire_pressure": [ 40.1, 39.9, | : [ 40.1,  |     3 |   data |
|       +skip                    | [ 40.1, 39.9, 37.7, 40.4 ] },  | 40.1, 39.9 |     3 |        |
|       skip                     | 40.1, 39.9, 37.7, 40.4 ] },    | , 39.9, 37 |     3 |        |
|       skip                     | , 39.9, 37.7, 40.4 ] },   { "m | 39.9, 37.7 |     3 |        |
|       skip                     | 39.9, 37.7, 40.4 ] },   { "mak | , 37.7, 40 |     3 |        |
|       skip                     | , 37.7, 40.4 ] },   { "make":  | 37.7, 40.4 |     3 |        |
|       skip                     | 37.7, 40.4 ] },   { "make": "K | , 40.4 ] } |     3 |        |
|       skip                     | , 40.4 ] },   { "make": "Kia", | 40.4 ] },  |     3 |        |
|       skip                     | 40.4 ] },   { "make": "Kia",   | ] },   { " |     3 |        |
|       -skip                    | ] },   { "make": "Kia",    "da | },   { "ma |     3 |        |
|     -object                    | },   { "make": "Kia",    "data | ,   { "mak |     2 |        |
|       no match                 | "make": "Toyota", "data": {"mo | : "Toyota" |     3 |   data |
|       skip                     | "Toyota", "data": {"model": "C | , "data":  |     3 |        |
|       match                    | "data": {"model": "Camry",  "y | : {"model" |     3 |   data |
|       +object                  | {"model": "Camry",  "year":"20 | "model": " |     3 |        |
|         no match               | "model": "Camry",  "year":"201 | : "Camry", |     4 |   year |
|         skip                   | "Camry",  "year":"2018", "rati | ,  "year": |     4 |        |
|         match                  | "year":"2018", "rating":"A"},  | :"2018", " |     4 |   year |    <===== (2)
|         string                 | "2018", "rating":"A"}, "tire_p | , "rating" |     4 |        |
|         string                 | "2018", "rating":"A"}, "tire_p | , "rating" |     4 |        |
```
I've annotated two lines with (1) and (2); these look the same to me as far as the buffer position, but it's a skip on (1) and a match on (2).  I guess it's because the depth is different.     The if-then-on-object's-key() approach does not have the skips, nor does 
```
	simdjson::ondemand::object data = o["data"];
	std::string_view model = data["model"];
	std::string_view year = data["year"];
	std::string_view rating = data["rating"];
```
I tried find_field() a little bit, but I couldn't quite get that to work.  

**But**: I've solved what I've needed to, I'll either use the key() approach or the approach directly above, using array-lookup from a predefined object.  Thanks!

P.S. IMHO, the logging is such a nice, easy to understand feature that it would be a great addition to the examples in the documentation!!

**Comment by lemire** [comment]
@drcdr Would you be willing to provide a short text explaining how you used the logging. It does not need to be very long... like ten lines. We would gladly add it to the documentation. Having the point of view of someone who used it to help themselves would be valuable. (Note: I really mean something short, not a whole essay.)


(Credit: The logging was created by @jkeiser.)

**Comment by drcdr** [comment]
_Sure, here is a high-level summary of my approach, leading into a very short tutorial as well; let me know if this works or if you'd like something different._

An on_demand parser can be designed in several different ways.   For more (historical) information, see (e.g.) [this (0.6.0)](https://simdjson.org/api/0.6.0/md_doc_ondemand.html), [this (0.8.0)](https://simdjson.org/api/0.8.0/md_doc_ondemand.html) and [this (2.0.0)](https://simdjson.org/api/2.0.0/md_doc_ondemand_design.html).  Additional information and context can also be found in the simdjson github issues and discussions regarding the on-demand parser.

To better understand the operation of your on-demand parser, and whether it is performing as well as you think it should be, there is a nice logger feature built in to simdjson!  To use it,
1)	Define the pre-processor directive SIMDJSON_VERBOSE_LOGGING, which enables logging in simdjson.h.
2)	In simdjson.h, be aware of the defines for LOG_EVENT_LEN, LOG_BUFFER_LEN, LOG_SMALL_BUFFER_LEN.  You may need to make them a little larger, depending on your application.
3)	Run your parser.  It may generate a lot of logging output; adding breaks that show each section may be helpful.

The log’s output will show step-by-step information on state, buffer pointer position, depth, and key retrieval status.

Example:
```
auto cars_json = R"( [
	{ "make": "Toyota", "data": {"model": "Camry",  "year":"2018", "rating":"A"}, "tire_pressure": [ 40.1, 39.9, 37.7, 40.4 ] },
] )"_padded;
static simdjson::ondemand::parser  parser;
simdjson::ondemand::document doc = parser.iterate(cars_json);

// Try one or the other of the following sections, to see the difference in the logging output.
{   	
	// ----- This section has a longer log -----
	printf("***** make *****\n");
	std::string_view make = o["make"];
	printf("***** data/model *****\n");
	std::string_view model = o["data"]["model"];
	printf("***** data/year *****\n");
	std::string_view year = o["data"]["year"];
}
{
	// ----- This section has a shorter log, it is more efficient. -----
	simdjson::ondemand::object data = o["data"];
	std::string_view model = data["model"];
	std::string_view year = data["year"];
	std::string_view rating = data["rating"];
}

```

**Comment by lemire** [comment]
Thanks. Let us work on including this in the documentation. Meanwhile, you may want to update to the latest version of simdjson (3.1.0), it should solve your problem above.

**Comment by lemire** [comment]
@drcdr
I am adding your content to our documentation.

Note that simdjson is on godbolt which allows you to explore the assembly output:

https://godbolt.org/z/xE4GWs573
